src/theory/quantifiers/sygus/cegis_core_connective.h

```diff
@@ -0,0 +1,393 @@
+    /** Add term n to pool whose sygus analog is s */
+    void addToPool(Node n, Node s);
+    /** Add a refinement point to this component */
+    void addRefinementPt(Node id, const std::vector<Node>& pt);
+    /** Add a false case to this component */
+    void addFalseCore(Node id, const std::vector<Node>& u);
+    /**
+     * Selects a node from passerts that evaluates to false on point mv if one
+     * exists, or otherwise returns false.
+     * The argument mvId is an identifier used for indexing the point mv.
+     * The argument asserts stores the current candidate solution (set D in
+     * Variant #2 described above). If the method returns true, it updates
+     * an (the node version of asserts) to be the conjunction of the nodes
+     * in asserts.
+     *
+     * If true is returned, it is removed from passerts.
+     */
+    bool addToAsserts(CegisCoreConnective* p,
+                      std::vector<Node>& passerts,
+                      const std::vector<Node>& mvs,
+                      Node mvId,
+                      std::vector<Node>& asserts,
+                      Node& an);
+
+    /**
+     * Get a refinement point that n evalutes to true on, taken from the
+     * d_refinementPt trie, and store it in ss. The set visited is the set
+     * of leaf nodes (reference by their data) that we have already processed
+     * and should be ignored.
+     */
+    Node getRefinementPt(CegisCoreConnective* p,
+                         Node n,
+                         std::unordered_set<Node, NodeHashFunction>& visited,
+                         std::vector<Node>& ss);
+    /** Get term pool, i.e. pool(A)/pool(B) in the algorithms above */
+    void getTermPool(std::vector<Node>& passerts) const;
+    /**
+     * Get the sygus solution corresponding to the Boolean connective for
+     * this component applied to conj. In particular, this returns a
+     * right-associative chain of applications of sygus constructor d_scons
+     * to the sygus analog of formulas in conj.
+     */
+    Node getSygusSolution(std::vector<Node>& conjs) const;
+    /** debug print summary (for debugging) */
+    void debugPrintSummary(std::ostream& os) const;
+
+   private:
+    /** The original formula for the pre/post condition A/B. */
+    Node d_this;
+    /**
+     * The sygus constructor for constructing solutions based on the core
+     * connective algorithm. This is a sygus datatype constructor that
+     * encodes applications of AND or OR.
+     */
+    Node d_scons;
+    /** This is pool(A)/pool(B) in the algorithms above */
+    std::vector<Node> d_cpool;
+    /**
+     * A map from the formulas in the above vector to their sygus analog.
+     */
+    std::map<Node, Node> d_cpoolToSol;
+    /**
+     * An index of list of predicates such that each list ( P1, ..., Pn )
+     * indexed by this trie is such that P1 ^ ... ^ Pn is equivalent to false.
+     * We call this a "false core".
```

Comment:
In the implementation you add uasserts to this collection. My understanding is that when you do this, uasserts corresponds to W of the algorithm. Is this
so? If it is, is d_falseCores actually cores(B)? if it is, can you mention this?

Reply:
Yes, ill add.